feat: added support for message provider

pulphix opened the pull request:
I did the porting to FastAPI, based on the work of @tuan-pham.
Please let me know if there are any areas of improvement.
Thank you.

elliottmurray commented on tests/test_message_provider.py:
```diff
@@ -0,0 +1,162 @@
+import os
+# import pytest
```
remove comment

pulphix replied:
Done :)

elliottmurray commented:
So this looks good but a bit to digest so give me a couple of days. There are some clean up (can you fix the semantic PR issue?) I'll point out as I do that review

tuan-pham commented on pact/http_proxy.py:
```diff
@@ -0,0 +1,57 @@
+"""Http Proxy to be used as provider url in verifier."""
```
The http_proxy looks good with FastAPI.👍 
 Thanks!

tuan-pham commented:
> I did the porting to FastAPI, based on the work of @tuan-pham.
> Please let me know if there are any areas of improvement.
> Thank you.

Take me sometime to review some of the code I wrote. But the http_proxy looks pretty good, thanks!

tuan-pham commented on tests/test_message_provider.py:
```diff
@@ -0,0 +1,162 @@
+        mock_start_proxy.assert_called_once()
+        mock_stop_proxy.assert_called_once()
+
+    # @pytest.mark.skip(reason="Caused BrokenPipeError in PactStartShutdownServerTestCase::test_stop_fails_posix. Need to investigate")
```
I think this caused by Flash. Now that it's replaced with FastAPI, I think we can remove the commented code

pulphix replied:
Done

tuan-pham commented on tests/test_message_provider.py:
```diff
@@ -0,0 +1,162 @@
+    def setUp(self):
+        super(StartProxyTestCase, self).setUp()
+
+    # @pytest.mark.skip(reason="Caused BrokenPipeError in PactStartShutdownServerTestCase::test_stop_fails_posix. Need to investigate")
```
Same here

pulphix replied:
Done

pulphix commented:
> So this looks good but a bit to digest so give me a couple of days. There are some clean up (can you fix the semantic PR issue?) I'll point out as I do that review

I changed the PR description to 
feat: added support for message provider
But I'm still seeing the Semantic Pull Request issue, any suggestions?

elliottmurray commented:
> > So this looks good but a bit to digest so give me a couple of days. There are some clean up (can you fix the semantic PR issue?) I'll point out as I do that review
> 
> I changed the PR description to
> feat: added support for message provider
> But I'm still seeing the Semantic Pull Request issue, any suggestions?

There is a security (and I think a wip) prefix commit message. That is not part of Semantic Messaging. 

https://github.com/zeke/semantic-pull-requests

elliottmurray commented:
So I'm doing some testing with the local examples and I think there is a small path issue trying to find the pacts on verify. I run pytest from the examples message directory.

```
self = <pact.verifier.Verifier object at 0x10ae7ec70>, enable_pending = False, include_wip_pacts_since = None
pacts = ('<PATH>/pact-python/examples/message/detectcontentlambda_message-documentservice_message.json',)
kwargs = {'verbose': False}
missing_files = ['<PATH>/pact-python/examples/message/detectcontentlambda_message-documentservice_message.json']

    def verify_pacts(self, *pacts, enable_pending=False, include_wip_pacts_since=None, **kwargs):
        """Verify our pacts from the provider.

        Returns:
          success: True if no failures
          logs: some tbd output of logs

        """
        self.validate_publish(**kwargs)

        missing_files = [path for path in pacts if not path_exists(path)]
        if missing_files:
>           raise Exception("Missing pact files {}".format(missing_files))
E           Exception: Missing pact files ['<PATH>/pact-python/examples/message/detectcontentlambda_message-documentservice_message.json']

../../pact/verifier.py:47: Exception
```
So I think 2 related path issues. The name of the file it is looking for last suffix is documentservice_message and the name you are generating is detectcontentlambda_message-contentprovider_message.json. I suspect on generating the file there is an issue with getting provider/consumer names mixed up?

Also you are generating this in the pacts folder (which I think is correct) but the verify test seems to be looking in root.

elliottmurray commented:
> So I'm doing some testing with the local examples and I think there is a small path issue trying to find the pacts on verify. I run pytest from the examples message directory.
> 
> ```
> self = <pact.verifier.Verifier object at 0x10ae7ec70>, enable_pending = False, include_wip_pacts_since = None
> pacts = ('<PATH>/pact-python/examples/message/detectcontentlambda_message-documentservice_message.json',)
> kwargs = {'verbose': False}
> missing_files = ['<PATH>/pact-python/examples/message/detectcontentlambda_message-documentservice_message.json']
> 
>     def verify_pacts(self, *pacts, enable_pending=False, include_wip_pacts_since=None, **kwargs):
>         """Verify our pacts from the provider.
> 
>         Returns:
>           success: True if no failures
>           logs: some tbd output of logs
> 
>         """
>         self.validate_publish(**kwargs)
> 
>         missing_files = [path for path in pacts if not path_exists(path)]
>         if missing_files:
> >           raise Exception("Missing pact files {}".format(missing_files))
> E           Exception: Missing pact files ['<PATH>/pact-python/examples/message/detectcontentlambda_message-documentservice_message.json']
> 
> ../../pact/verifier.py:47: Exception
> ```
> 
> So I think 2 related path issues. The name of the file it is looking for last suffix is documentservice_message and the name you are generating is detectcontentlambda_message-contentprovider_message.json. I suspect on generating the file there is an issue with getting provider/consumer names mixed up?
> 
> Also you are generating this in the pacts folder (which I think is correct) but the verify test seems to be looking in root.

Actually think it's a bit more than that. I changed in test_message_provider.py the provider name to match the ones in the consumer tests. Then moved the generated file to the correct place and re-run. It's still failing. I think there is a mismatch in the contracts somewhere. 

So how I've done it in others is to have a contrived example where I rely on pytest to run my consumer tests first then my provider verifies against that json output. I think this is what is supposed to be happening here but I'm not 100% sure.

elliottmurray commented:
Also might be worth getting a rebase from master. This branch is quite old and while the changes are small there is a test that breaks on Mac for me I've since fixed.

pulphix commented:
> Also might be worth getting a rebase from the master. This branch is quite old and while the changes are small there is a test that breaks on Mac for me I've since fixed.

@elliottmurray 
I'm working in a new branch rebased on master, I have fixed the example tests.
I have a question for you, there is any specific reason with we append _message on the name of the provider and consumer?
This is not going to alter the behavior with the pact broker?

pulphix commented:
Created a new pull request
